### PR TITLE
feat: :recycle: Implemented code to handle generic commands

### PR DESCRIPTION
**Pull request by EricRootLee**
*Replace this paragraph with a description of what this PR is changing or adding, and why.*

### List of issues which are fixed by the PR
*You must list at least one issue.*

- `--help` shows available commands
- `<command> --help` shows the arguments of a specific command
- implement     `publish` and `update` to create the `commands` folder
- implement `WidgetbookCommand` similar to `MasonCommand
- implement `WidgetbookCommandRunner` similar to `MasonCommandRunner
- [build GitHub action for the test suite](https://app.clickup.com/t/3194g2k)
- [build a test suite for generic commands (see how mason does this)](https://app.clickup.com/t/3194f17) (Almost done)

### Screenshots
*If applicable, add screenshots to help explain the changes.*

### Checklist

- [ ] I signed the [CLA].
- [ ] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] I added new tests to check the change I am making].
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on [Discord].

[CLA]: https://docs.google.com/forms/d/e/1FAIpQLScuRfjUzENsLsmQgqZlGLxMKbFi7zuXoPARyXytoyQrq7ntUw/viewform
[Discord]: https://discord.com/invite/zT4AMStAJA

### CODE REVIEW

File: packages/widgetbook_cli/bin/command/analyze.dart [hunk 1]
```diff
@@ -0,0 +1,201 @@
+import 'dart:io';
+
+import 'package:interact/interact.dart';
+import 'package:mason_logger/mason_logger.dart';
+import 'package:pubspec/pubspec.dart';
```
Comment (jenshor) [inline packages/widgetbook_cli/bin/command/analyze.dart:5]: Good that you've looked for a package. How maintained is this?

File: packages/widgetbook_cli/bin/command/publish.dart [hunk 2]
```diff
@@ -0,0 +1,234 @@
+import 'dart:io';
+
+import 'package:ci/ci.dart' as ci;
```
Comment (jenshor) [inline packages/widgetbook_cli/bin/command/publish.dart:3]: ooooohhh, cool that there's a package for this 👍🏻
Great find!

File: packages/widgetbook_cli/bin/command/publish.dart [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+  }
+
+  Future<String> _getActorName() async {
+    final results = await Process.run('git', ['config', 'user.name']);
+    return results.stdout as String;
+  }
+
+  Future<String> _getRepositoryName() async {
+    final results = await Process.run('git', ['rev-parse', '--show-toplevel']);
+    return results.stdout.toString().split('/').last;
```
Comment (jenshor) [inline packages/widgetbook_cli/bin/command/publish.dart:232]: should use git package

File: packages/widgetbook_cli/bin/command/publish.dart [hunk 4]
```diff
@@ -0,0 +1,234 @@
...
+      }
+    } else {
+      logger.err('Could not create .zip file for upload.');
+    }
+    return ExitCode.success.code;
+  }
+
+  Future<String> _getActorName() async {
+    final results = await Process.run('git', ['config', 'user.name']);
+    return results.stdout as String;
```
Comment (jenshor) [inline packages/widgetbook_cli/bin/command/publish.dart:227]: should use git package

File: packages/widgetbook_cli/bin/command/publish.dart [hunk 5]
```diff
@@ -0,0 +1,234 @@
...
+
+    if (ci.isCI && ci.Vendor.IS_GITHUB_ACTIONS) {
+      repository = Platform.environment['GITHUB_ACTION_REPOSITORY'];
+
+      actor = Platform.environment['GITHUB_ACTOR'];
+    } else {
+      repository =
+          results['repository'] as String? ?? await _getRepositoryName();
+      actor = results['actor'] as String? ?? await _getActorName();
+    }
```
Comment (jenshor) [inline packages/widgetbook_cli/bin/command/publish.dart:125]: Just to understand how this works: how is it determined if a CI is e.g. Github Action?

File: packages/widgetbook_cli/bin/command/upgrade.dart [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+class UpgradeCommand extends WidgetbookCommand {
+  UpgradeCommand({
+    required PubUpdater pubUpdater,
+    super.logger,
+  }) : _pubUpdater = pubUpdater;
+
+  final PubUpdater _pubUpdater;
+
+  @override
+  final String description = 'Upgrade widgetbook cli';
```
Comment (jenshor) [inline packages/widgetbook_cli/bin/command/upgrade.dart:17]: "Upgrade Widgetbook CLI"

File: packages/widgetbook_cli/bin/main.dart [hunk 7]
```diff
@@ -1,209 +1,15 @@
...
-        'HINT: No pull-request information available. Therefore, no review will'
-        ' be created. See docs for more information.',
-      );
-    }
-  } else {
-    print('Could not create .zip file for upload.');
-  }
+Future flushThenExit(int status) {
+  return Future.wait<void>([stdout.close(), stderr.close()])
+      .then<void>((_) => exit(status));
```
Comment (jenshor) [inline packages/widgetbook_cli/bin/main.dart:14]: I'm assuming that we need this code snippet, but I don't know why. 
Can you document why this method exists?

File: packages/widgetbook_cli/bin/version.dart [hunk 8]
```diff
@@ -1,3 +1 @@
-import 'package:version/version.dart';
-
-Version currentVersion = Version(0, 1, 0);
+const String currentVersion = '0.1.0-dev.35';
```
Comment (jenshor) [inline packages/widgetbook_cli/bin/version.dart:1]: This is not the correct version which makes me doubt that the update command is properly implemented.

File: packages/widgetbook_cli/bin/widgetbook_command_runner.dart [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+import 'command/logout.dart';
+import 'command/publish.dart';
+import 'version.dart';
+
+/// The package name.
+const packageName = 'widgetbook_cli';
+
+class WidgetbookCommandRunner extends CommandRunner<int> {
+  WidgetbookCommandRunner({
+    Logger? logger,
```
Comment (jenshor) [inline packages/widgetbook_cli/bin/widgetbook_command_runner.dart:21]: Do we need a license featuring Angelov in this file?

File: packages/widgetbook_cli/pubspec.yaml [hunk 10]
```diff
@@ -14,18 +14,25 @@ dev_dependencies:
   freezed: ^2.0.3
   json_serializable: ^6.2.0
   lints: ^1.0.0
+  mocktail: ^0.3.0
   very_good_analysis: ^3.0.0
 dependencies: 
   archive: ^3.3.0
   args: ^2.3.0
+  ci: ^0.1.0
```
Comment (jenshor) [inline packages/widgetbook_cli/pubspec.yaml:22]: looks good to me 👍🏻

File: packages/widgetbook_cli/pubspec.yaml [hunk 11]
```diff
@@ -14,18 +14,25 @@ dev_dependencies:
...
   dio: ^4.0.6
   file: ^6.1.2
   freezed_annotation: ^2.0.3
-  widgetbook_git: ^2.1.0 
   http_parser: ^4.0.1
+  interact: ^2.1.1
```
Comment (jenshor) [inline packages/widgetbook_cli/pubspec.yaml:27]: not sure if we need this

File: packages/widgetbook_cli/pubspec.yaml [hunk 12]
```diff
@@ -14,18 +14,25 @@ dev_dependencies:
...
+  mason_logger: ^0.1.3
   path: ^1.8.2
+  pub_updater: ^0.2.2
```
Comment (jenshor) [inline packages/widgetbook_cli/pubspec.yaml:30]: looks good to me

File: packages/widgetbook_cli/pubspec.yaml [hunk 13]
```diff
@@ -14,18 +14,25 @@ dev_dependencies:
...
   dio: ^4.0.6
   file: ^6.1.2
   freezed_annotation: ^2.0.3
-  widgetbook_git: ^2.1.0 
   http_parser: ^4.0.1
+  interact: ^2.1.1
+  mason_logger: ^0.1.3
   path: ^1.8.2
+  pub_updater: ^0.2.2 
+  pubspec: ^2.3.0
```
Comment (jenshor) [inline packages/widgetbook_cli/pubspec.yaml:31]: looks ok to me, not sure if this package is well maintained

Comment (EricRootLee) [reply packages/widgetbook_cli/pubspec.yaml:31]: Yes seems like ,  last  review was
Published [3 months ago](https://pub.dev/packages/pubspec)

File: packages/widgetbook_cli/bin/command/publish.dart (hunk 5 above)
Comment (EricRootLee) [reply packages/widgetbook_cli/bin/command/publish.dart:125]: the packages gets info from Platform Environment  checks is Github  Actions is the current CI vendor. `Platform.environment.containsKey('GITHUB_ACTIONS');`

File: packages/widgetbook_cli/bin/command/analyze.dart (hunk 1 above)
Comment (EricRootLee) [reply packages/widgetbook_cli/bin/command/analyze.dart:5]: Yes seems like , last review was
Published [3 months ago](https://pub.dev/packages/pubspec)

File: packages/widgetbook_cli/bin/widgetbook_command_runner.dart (hunk 9 above)
Comment (EricRootLee) [reply packages/widgetbook_cli/bin/widgetbook_command_runner.dart:21]: I think so; most of the code is imported from the mason packages. Trying to figure out how I'll do the citation

File: packages/widgetbook_cli/bin/version.dart (hunk 8 above)
Comment (EricRootLee) [reply packages/widgetbook_cli/bin/version.dart:1]: I left this there deliberately to test if the update works.

File: packages/widgetbook_cli/bin/cli_debug.dart [hunk 14]
```diff
@@ -3,5 +3,5 @@ import 'main.dart' as cli;
 
 void main(List<String> args) {
   Flavor().strategy = DeploymentStrategy.debug;
-  cli.main(args);
+  cli.main(args, 'widgetbook_debug');
```
[the remark this answers is not in the document]
Comment (EricRootLee) [reply packages/widgetbook_cli/bin/cli_debug.dart:6]: When no command is passed,   args is an empty list.
 The reason that I am passing that parameter is for the update statement to be in sync with the actual flavor of the ci that someone is running `Run [widgetbook_debug ,widgetbook_staging ,widgetbook] update to update.`

File: packages/widgetbook_cli/bin/cli_production.dart [hunk 15]
```diff
@@ -3,5 +3,5 @@ import 'main.dart' as cli;
 
 void main(List<String> args) {
   Flavor().strategy = DeploymentStrategy.production;
-  cli.main(args);
+  cli.main(args, 'widgetbook_production');
```
[the remark this answers is not in the document]
Comment (EricRootLee) [reply packages/widgetbook_cli/bin/cli_production.dart:6]: updated `widgetbook_production' to `widgetbook

When no command is passed,   args is an empty list.
 The reason that I am passing that parameter is for the update statement to be in sync with the actual flavor of the ci that someone is running `Run [widgetbook_debug ,widgetbook_staging ,widgetbook] update to update.`

File: packages/widgetbook_cli/bin/cli_staging.dart [hunk 16]
```diff
@@ -3,5 +3,5 @@ import 'main.dart' as cli;
 
 void main(List<String> args) {
   Flavor().strategy = DeploymentStrategy.staging;
-  cli.main(args);
+  cli.main(args, 'widgetbook_staging');
```
[the remark this answers is not in the document]
Comment (EricRootLee) [reply packages/widgetbook_cli/bin/cli_staging.dart:6]: When no command is passed, args is an empty list.
The reason that I am passing that parameter is for the update statement to be in sync with the actual flavor of the ci that someone is running Run [widgetbook_debug ,widgetbook_staging ,widgetbook] update to update.

File: packages/widgetbook_cli/bin/command/analyze.dart (hunk 1 above)
Comment (jenshor) [reply packages/widgetbook_cli/bin/command/analyze.dart:5]: Can you check open issues? But sounds fine

File: packages/widgetbook_cli/bin/command/publish.dart (hunk 5 above)
Comment (jenshor) [reply packages/widgetbook_cli/bin/command/publish.dart:125]: It's what I've expected 👍🏻

File: packages/widgetbook_cli/bin/version.dart (hunk 8 above)
Comment (jenshor) [reply packages/widgetbook_cli/bin/version.dart:1]: Okay. But in the final release this value has to be correct 👍🏻

**Review by jenshor — CHANGES_REQUESTED** [review]

File: packages/widgetbook_cli/bin/command/publish.dart (hunk 3 above)
Comment (EricRootLee) [reply packages/widgetbook_cli/bin/command/publish.dart:232]: Yes already solved this .. now using widgetbook_git

File: packages/widgetbook_cli/bin/command/publish.dart (hunk 4 above)
Comment (EricRootLee) [reply packages/widgetbook_cli/bin/command/publish.dart:227]: Yes already solved this .. now using widgetbook_git

File: packages/widgetbook_cli/bin/command/analyze.dart (hunk 1 above)
Comment (EricRootLee) [reply packages/widgetbook_cli/bin/command/analyze.dart:5]: Yes I did check , there is only 2 issues  at the ,moment , nothing major that can affect  the usage and implementation in the cli

**Comment by EricRootLee** [comment]
This pr might be very painful to review, it has merges from the previous 2 prs , sorry about that

File: packages/widgetbook_cli/bin/commands/setup.dart [hunk 17]
```diff
@@ -0,0 +1,133 @@
...
+
+    final projectName =
+        systemTempDir.path.split('/')[systemTempDir.path.split('/').length - 2];
+
+    publishProgress.complete('Analyzing Project done');
+
+    logger.success('Project Name : '
+        '$projectName');
+
+    dartPackages.removeWhere((item) => item.projectName == 'packages');
```
Comment (jenshor) [inline packages/widgetbook_cli/bin/commands/setup.dart:39]: why is this line necessary?

File: packages/widgetbook_cli/bin/commands/setup.dart [hunk 18]
```diff
@@ -0,0 +1,133 @@
...
+      logger.info('No Flutter Projects Found\n');
+      return ExitCode.noInput.code;
+    }
+    if (dartPackages.isNotEmpty) {
+      logger.info('Found ${dartPackages.length} other Dart Package(s) \n');
+      for (final i in dartPackages) {
+        logger.info('> ${i.projectName}');
+      }
+      logger.info('\n');
+    }
```
Comment (jenshor) [inline packages/widgetbook_cli/bin/commands/setup.dart:59]: This code can be optimized.

File: packages/widgetbook_cli/bin/commands/upgrade.dart [hunk 19]
```diff
@@ -0,0 +1,76 @@
+// The MIT License (MIT)
+// Copyright (c) 2022 Felix Angelov
+// Copyright (c) 2022 Widgetbook GmbH
```
Comment (jenshor) [inline packages/widgetbook_cli/bin/commands/upgrade.dart:3]: Widgetbook on top

File: packages/widgetbook_cli/bin/main.dart [hunk 20]
```diff
@@ -1,209 +1,43 @@
...
-  final branch = args['branch'] as String? ?? currentBranch.branchName;
-  final repository = args['repository'] as String;
-  final commit = args['commit'] as String? ?? currentBranch.sha;
-  final actor = args['actor'] as String;
-  final gitProvider = args['git-provider'] as String;
-  final gitHubToken = args['github-token'] as String?;
-  final prNumber = args['pr'] as String?;
+// The MIT License (MIT)
+// Copyright (c) 2022 Felix Angelov
+// Copyright (c) 2022 Widgetbook GmbH
```
Comment (jenshor) [inline packages/widgetbook_cli/bin/main.dart:3]: Widgetbook on top

Comment (jenshor) [reply packages/widgetbook_cli/bin/main.dart:3]: Do we really need the License here?

File: packages/widgetbook_cli/bin/models/project.dart [hunk 21]
```diff
@@ -0,0 +1,32 @@
+import '../commands/helpers/helpers.dart';
+
+class Project {
```
Comment (jenshor) [inline packages/widgetbook_cli/bin/models/project.dart:3]: This should be a freezed class. 
The name is suboptimal because we also have a `Project` in Widgetbook Cloud which will cause naming conflicts in the future. `Package` or `DartPackage` might be a lot better.

File: packages/widgetbook_cli/bin/parsers/exceptions.dart [hunk 22]
```diff
@@ -0,0 +1,21 @@
+import 'package:widgetbook_exception/widgetbook_exception.dart';
+
+class WidgetbookDeployFailure extends WidgetbookException {
+  WidgetbookDeployFailure(super.message);
```
Comment (jenshor) [inline packages/widgetbook_cli/bin/parsers/exceptions.dart:4]: All these exceptions should have a pre-defined message. Otherwise, we could also make `WidgetbookException` non abstract and remove this code.

File: packages/widgetbook_cli/bin/widgetbook_command_runner.dart [hunk 23]
```diff
@@ -0,0 +1,141 @@
+// The MIT License (MIT)
+// Copyright (c) 2022 Felix Angelov
+// Copyright (c) 2022 Widgetbook GmbH
```
Comment (jenshor) [inline packages/widgetbook_cli/bin/widgetbook_command_runner.dart:3]: Widgetbook on top

File: packages/widgetbook_cli/bin/widgetbook_command_runner.dart [hunk 24]
```diff
@@ -0,0 +1,141 @@
...
+      'version',
+      negatable: false,
+      help: 'Print the current version.',
+    );
+    addCommand(UpgradeCommand(logger: _logger, pubUpdater: _pubUpdater));
+    addCommand(LoginCommand(logger: _logger));
+    addCommand(LogoutCommand(logger: _logger));
+    addCommand(PublishCommand(logger: _logger));
+    addCommand(SetupCommand(logger: _logger));
+    addCommand(AnalyzeCommand(logger: _logger));
```
Comment (jenshor) [inline packages/widgetbook_cli/bin/widgetbook_command_runner.dart:56]: This should only contain the commands which actually work and are tested.

File: packages/widgetbook_git/lib/src/git_dir.dart [hunk 25]
```diff
@@ -123,6 +124,12 @@ class GitDir {
         .toBranchReference();
   }
 
+  Future<int> unCommitedChangesCount() async {
+    final changes = await runCommand(const ['status', '--porcelain']);
+    final lines = const LineSplitter().convert(changes.stdout.toString());
+    return lines.length;
+  }
+
```
Comment (jenshor) [inline packages/widgetbook_git/lib/src/git_dir.dart:132]: I think we should undo this change.

**Review by jenshor — CHANGES_REQUESTED** [review]
We should merge this first which will simplify the review of the other PR.

File: packages/widgetbook_git/lib/src/git_dir.dart (hunk 25 above)
Comment (EricRootLee) [reply packages/widgetbook_git/lib/src/git_dir.dart:132]: This was already fixed

File: packages/widgetbook_cli/bin/commands/publish.dart [hunk 26]
```diff
@@ -0,0 +1,313 @@
...
+    final publishProgress = logger.progress(
+      'Uploading build',
+    );
+    void publishProgressFailure() => publishProgress.fail();
+    final path = results['path'] as String;
+
+    if (!await GitDir.isGitDir(path)) {
+      publishProgressFailure();
+      usageException('Directory from "path" is not a Git folder');
+    }
```
Comment (jenshor) [inline packages/widgetbook_cli/bin/commands/publish.dart:125]: This error handling looks weird on the first look

Comment (jenshor) [reply packages/widgetbook_cli/bin/commands/publish.dart:125]: Ok, it's fine

File: packages/widgetbook_cli/bin/parsers/exceptions.dart [hunk 27]
```diff
@@ -0,0 +1,76 @@
+import 'package:widgetbook_exception/widgetbook_exception.dart';
+
+class WidgetbookDeployFailure implements WidgetbookException {
```
Comment (jenshor) [inline packages/widgetbook_cli/bin/parsers/exceptions.dart:3]: How is the naming convention for MasonCLI? To me, it looks odd that a `WidgetbookDeployFailure` implements a `WidgetbookException`. 

A failure is not an exception.

File: packages/widgetbook_cli/bin/commands/publish.dart (hunk 26 above)
Comment (EricRootLee) [reply packages/widgetbook_cli/bin/commands/publish.dart:125]: Hi, aha the  `!await GitDir.isGitDir(path)` returns a bool
. If the path is not a git folder (false) `usageException()` throws a Usage Exception. it is an internal handling error from `args`.

File: packages/widgetbook_git/lib/src/git_dir.dart [hunk 28]
```diff
@@ -123,6 +123,12 @@ class GitDir {
         .toBranchReference();
   }
 
+  Future<String> unCommitedChanges() async {
+    final changes = await runCommand(const ['status']);
+
+    return changes.stdout.toString();
+  }
+
```
Comment (jenshor) [inline packages/widgetbook_git/lib/src/git_dir.dart:131]: Do we need these changes?
